src/participant.rs

```diff
@@ -168,20 +168,26 @@ where
     }
 }
 
-pub(crate) trait ProtocolParticipant {
-    /// Input type used to process incoming messages.
-    type Input;
+/// These are the public-facing methods that must be implemented for a given
+/// protocol.
+pub trait PublicProtocolParticipant {
```

Comment:
Minor, but I'd prefer a renaming of `PublicProtocolParticipant` -> `ProtocolParticipant` and `ProtocolParticipant` -> `InnerProtocolParticipant` (or something like that). The reasoning is that the public API only exposes `PublicProtocolParticipant` so the `Public` seems redundant.